### PR TITLE
Feat: Structured outputs

### PR DESCRIPTION
**Pull request by arthurhjorth**
This is my first contribution to a larger project, so I'm writing this with trepidation.

### Description
This PR adds support for structured outputs on the `mlx-lm.server` and for `BatchGenerator`. Structured outputs are achieved by creating a stateful logits_processor from a JSON schema using `outlines`'s `OutlinesCoreLogitsProcessor`. It keeps track of and forwards its own state for each token, so it works out of the box for how logits_processors are handled by `mlx-lm` and so aligns nicely with the recent additions of logits_processors to `BatchGenerator` (v 0.30.1).

### Changes
* `generate.py`:  `batch_generate()` accepted `logits_processors` as an argument, but never passed them on to `BatchGenerator`. So I made one tiny change in line 1299 to pass them in alongside prompts, tokens and caches.
* `structured.py` is where the `outlines` related stuff lives. There is some overhead when building each for the first time, so it is built around a cache of `OutlinesCoreLogitsProcessor` objects can be retrieved by tokenizer+schema or tokenizer+Type[BaseModel].
* `server.py`: Request handling now extracts JSON schema from requests, and if present gets a logit processor from the cache and passes that into generate calls.

Edit: by the way, I wrote pseudocode for the logits processors cache, but had an LLM write it for me. I then moved things around bit and added data classes since the LLMs code resulted in incomprehensible type hinting (List[Dict[str, List[int]...) etc. Just for full disclosure.

### Motivation
Structured outputs are fundamental for reliably working programatically with LLM outputs. For that reason, they're in demand for LLM-servers. My personal motivation for including it in batching is that most of my research focuses on text analyses of massive corpora, and I need to be able to specify precisely what I want the LLM to analyze/look for/label. Structured outputs is basically a requirement for that.

###  Final thoughts/comments:
Outlines has some issues. The current version of `outlines` installs an `outlines-core` that fails on fp16 models (#814). The only way to make it work is to force an update on only `outlines-core`, which makes it _work_ - but gives a pip error. I could see why you wouldn't want to include that as a requirement for `mlx-lm`. Maybe they will fix this, but as you see in #814, @awni [raised this with the outlines team four months ago](https://github.com/dottxt-ai/outlines/issues/1769), and it's still not fixed. Not super confidence inspiring.

The _minimal_ change would be to just keep the change to `generate` so logits_processors are at least passed into `BatchGenerator`. It feels like it might have been the intention anyway, since `logits_processors` are not used for anything else in there.

If you decide to go with the minimal change, I'm wondering if I should do another PR that instead focuses on aligning the ways in which logits_processors are passed into different kinds of generation functions, e.g. for `generate()` and `stream_generate()` you have to pass them in as kwargs:
```
g = generate(model, tokenizer, prompts[0], **{'logits_processors': logits_processors[0]})
```
whereas `batch_generate()` accept them as named args,

```
result = batch_generate(
    model, tokenizer, prompts, verbose=False, return_prompt_caches=True, logits_processors=logits_processors
)
```

### CODE REVIEW

**Comment by awni** [comment]
Indeed I don't think it makes sense to take outlines as a dependency here. But I would like to get the fix to the logits processors you made merged. Would you mind either sending that as a separate PR or removing the rest here?

**Comment by arthurhjorth** [comment]
Weekend had started in Denmark and was away for the weekend, sorry!

Very happy to - I'll make a separate PR based on a fresh pull tomorrow and submit that.

And re. outlines dependency, totally understandable. I'll just keep the server in a repo of its own in case someone wants to use that. It might still be useful to some.
